### PR TITLE
Add pre-conditioning matrix to Barker proposal

### PR DESCRIPTION
**Pull request by ismael-mendoza**
# Overview

This PR attempts to add a preconditioning matrix to the barker proposal implementation, as in appendix G of https://arxiv.org/abs/1908.11812

- Relevant discussion #723 
- Relevant issue #725 

# Discussion / Questions
- I changed the `_compute_acceptance_probability` to be flat given how it was necessary to matrix-multiply with the pre-conditioning matrix. Is that OK? 
- Do I need to update the PDF?  (not used anywhere in the codebase)
- It's not very clear in the Barker paper but the step_size (sigma) is a 'global scale' and the inverse mass matrix is separate.  See implementation from authors here (in R) - https://github.com/gzanella/barker/blob/master/functions.R

# Thank you for opening a PR!

 A few important guidelines and requirements before we can merge your PR:

 - [x] *If I add a new sampler*, there is an issue discussing it already;
 - [x] We should be able to understand what the PR does from its title only;
 - [x] There is a high-level description of the changes;
 - [x] There are links to *all* the relevant issues, discussions and PRs;
 - [x] The branch is rebased on the latest `main` commit;
 - [x] Commit messages follow these [guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html);
 - [x] The code respects the current naming conventions;
 - [x] Docstrings follow the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html)
 - [x] `pre-commit` is installed and configured on your machine, and you ran it before opening the PR;
 - [x] There are tests covering the changes;
 - [x] The doc is up-to-date;
 - [x] *If I add a new sampler** I added/updated related [examples](https://github.com/blackjax-devs/blackjax/tree/main/examples)

Consider opening a **Draft PR** if your work is still in progress but you would like some feedback from other contributors.

### CODE REVIEW

**Comment by ismael-mendoza** [comment]
@AdrienCorenflos I just added a test to specifically check the new implementation with the mass matrix and fixed all the tests that use Barker. I'm open to suggestions on the new test or on any other changes. Thank you!

**Comment by AdrienCorenflos** [comment]
Thanks for this. I think I now know how we can use general `Metrics` here. This requires a small addition to the `Metric` interface, whereby a metric would now also have a `scale` function argument, with signature
```python
# Callable[[ArrayLikeTree, Arr], ArrayLikeTree]
def scale(position: ArrayLikeTree, vector: ArrayLikeTree, inv=False) -> ArrayLikeTree:
    mass_matrix = mass_matrix_fn(position)
    # ravel everything etc
    if inv:
        return triangular_solve(mass_matrix_sqrt, vector)  # plus unravelling of course
    return mass_matrix_sqrt @ vector  # plus unravelling of course
```

@junpenglao would that be ok to add a function like this (essentially scaling a vector by the `inv_sqrt_mass` defined by the metric at position x).

If this is ok, then we can implement Barker's proposal with any sort of metric, even Riemannian, for cheap: 
https://github.com/ismael-mendoza/blackjax/blob/a61af36ab77dc9926ce2b54cdb3d25f6ecb05dd4/blackjax/mcmc/barker.py#L93-L95 would become
```python 
z = metric.scale(y, y, True) - metric.scale(x, x, True)  # Be careful here, it may be metric.scale(x, y, True) - metric.scale(y, x, True), need to check the detailed balanced condition properly and same below.
c_x = metric.scale(x, log_x)
c_y = metric.scale(y, log_y)
```
And something similar here
https://github.com/ismael-mendoza/blackjax/blob/a61af36ab77dc9926ce2b54cdb3d25f6ecb05dd4/blackjax/mcmc/barker.py#L245-L253

There would be some bookkeeping about shapes and all but these seem to be the only change.

@ismael-mendoza @junpenglao what do you think? It;s

**Comment by AdrienCorenflos** [comment]
This is not even covered in the papers by Livingstone and Zanella: it's a good example of why Blackjax composability is cool :D

File: tests/mcmc/test_barker.py
Comment (AdrienCorenflos) [inline tests/mcmc/test_barker.py:125]: Does this pass with higher precision -> this should really hold up to numerical precision.

File: blackjax/mcmc/barker.py
Comment (AdrienCorenflos) [inline blackjax/mcmc/barker.py:109]: Breaking change, this needs to default to previous behaviour.

File: blackjax/mcmc/barker.py
Comment (AdrienCorenflos) [inline blackjax/mcmc/barker.py:285]: Duplicated code, hopefully if we push for the use of the Metric, this will fully live outside this file.

File: tests/mcmc/test_barker.py
Comment (ismael-mendoza) [reply tests/mcmc/test_barker.py:125]: I will check, you are right

**Comment by ismael-mendoza** [comment]
Thanks @AdrienCorenflos that seems like a good idea to me as it avoids the code redundancy and allows us to use Riemannian metrics for free like you pointed out. 

On your comment here 
```
z = metric.scale(y, y, True) - metric.scale(x, x, True)  # Be careful here, it may be metric.scale(x, y, True) - metric.scale(y, x, True), need to check the detailed balanced condition properly and same below.
```

do you have suggestions for how to check the correct detailed balanced condition? I'm not too familiar with this and I assume the Barker paper wouldn't mention it as they only show the algorithm for a fixed mass matrix. Thanks!

**Comment by ismael-mendoza** [comment]
One quick follow up - am I missing something or would you need to implementations of `.scale` ? one for the `gaussian_euclidean` and one for the `gaussian_riemannian` metric?  The former one won't take the position?

**Comment by AdrienCorenflos** [comment]
> One quick follow up - am I missing something or would you need to implementations of `.scale` ? one for the `gaussian_euclidean` and one for the `gaussian_riemannian` metric? The former one won't take the position?


We do need this (or similar) to be added I think, but let's wait for @junpenglao he will likely have a better opinion than me on the right API/design choice for these as he has thought about it a lot more than I have.

**Comment by junpenglao** [comment]
> > One quick follow up - am I missing something or would you need to implementations of `.scale` ? one for the `gaussian_euclidean` and one for the `gaussian_riemannian` metric? The former one won't take the position?
> 
> We do need this (or similar) to be added I think, but let's wait for @junpenglao he will likely have a better opinion than me on the right API/design choice for these as he has thought about it a lot more than I have.

Yes sounds like a great addition to `metric`, the API makes sense as well.

**Comment by ismael-mendoza** [comment]
Hello @AdrienCorenflos and @junpenglao, 

Thanks @AdrienCorenflos for implemeting the metric scaling. Now I can proceed finishing this PR, I just had two remaining questions). 

First, I noticed section 6.1 of the barker [paper](https://arxiv.org/pdf/1908.11812) that the authors use the proposed adaptation of Andrieu and Thoms (2008) for their experiments. I think it would be nice to have the full version of their algorithm blackjax and I wondering if this adaptation has already been implemented somewhere for blackjax? or maybe there is already some clear alternative that should be used to adapt hmc algorithms? It's unclear to me if the `window_adaptation` function that is used commonly in NUTS would be appropriate here for example. 

Second, @AdrienCorenflos I was just wondering if you finished the derivation above and if there is some additional term that must be added? 

Thank you both for your help!

**Comment by AdrienCorenflos** [comment]
Oh right, I had completely forgotten about this bit. I'm sure I've got the derivation on a piece of paper *somewhere* 😬
Let me check tomorrow morning.

**Comment by AdrienCorenflos** [comment]
Regarding the adaptation bit, let's check when that's done

**Comment by AdrienCorenflos** [comment]
Actually, I think the easiest is if you implement Barker for the standard fixed mass matrix using this new scale thing and then I'll adapt it to the manifold case

**Comment by ismael-mendoza** [comment]
Hi @AdrienCorenflos I noticed something strange in the `_format_covariance` fnc that tripped me up when adding the scaling: 

```
if is_inv:
    inv_cov_sqrt = jscipy.linalg.cholesky(cov, lower=True)
    cov_sqrt = jscipy.linalg.solve_triangular(
        inv_cov_sqrt, identity, lower=True, trans=True
    )
else:
    cov_sqrt = jscipy.linalg.cholesky(cov, lower=False).T
    inv_cov_sqrt = jscipy.linalg.solve_triangular(
        cov_sqrt, identity, lower=True, trans=True
    )
```
am I misunderstanding or the `inv_cov_sqrt` will be transposed in the case that `is_inv = False` and not tranposed in the case that `is_inv=True`? Is that the desired implementation? I was just surprised that it would be inconsistent like that but maybe it has to do with other algorithms?

**Comment by AdrienCorenflos** [comment]
Yeah we have been discussing this with @junpenglao 
It's because the standard Euclidean metric passes in the inverse mass matrix, whereas the Riemannian uses the normal mass matrix... It's a bit inconsistent and we were going to look into making this coherent after this specific PR. For the time being, I'd recommend you make it work for the Euclidean in the new metric API and I'll take care of the Riemannian.

**Comment by ismael-mendoza** [comment]
@AdrienCorenflos it required some finagling from what I said above but I think I was able to update my PR with the new metric.scale that at least works with the Euclidean metric. I also needed to add an option to transpose the matrix sqrt when scaling since the Barker algorithm required both (tranposed and untransposed) as far as I could tell. I added it to the Eucliden, let me know if you want me to propagate to the Riemannian. 

Please let me know if you have any other suggestions.

File: blackjax/mcmc/barker.py
Comment (ismael-mendoza) [reply blackjax/mcmc/barker.py]: resolved in last commit, removed this duplicated code

File: blackjax/mcmc/barker.py
Comment (ismael-mendoza) [reply blackjax/mcmc/barker.py]: made the `inverse_mass_matrix` an optional argument in altest commit

File: tests/mcmc/test_barker.py
Comment (ismael-mendoza) [reply tests/mcmc/test_barker.py]: removed the tolerance in the latest commit so now it's default jax.numpy tolerance

**Comment by AdrienCorenflos** [comment]
Yes please, APIs should mirror each other.

**Comment by ismael-mendoza** [comment]
@AdrienCorenflos added it there as well

**Comment by AdrienCorenflos** [comment]
It all seems to work for the standard pre-conditioning, but I'm seeing a lack of invariance for the Riemannian metric:

image

I'll fix, FYI, it comes from the fact that the proposal is not symmetric in the Gaussian anymore, so you have to compute the pdf

**Review by AdrienCorenflos — COMMENTED** [review]
I've made changes so that the proposal works with Riemannian metrics.
There is one last thing to fix and then we're good and I'll leave it to you @ismael-mendoza  

-> the barker_nd sampling function was meant to take fully flat vectors, and here you pass it the metric which acts on trees. You are not seeing issues because your tests do not use nested trees, but this will fail. Can you modify the logic so that it simply does everything using metric on the non-flat vectors? This may require changing a few tests on the function here and there but should be easy

**Review by AdrienCorenflos — CHANGES_REQUESTED** [review]
I've made changes so that the proposal works with Riemannian metrics.
There is one last thing to fix and then we're good and I'll leave it to you @ismael-mendoza  

-> the barker_nd sampling function was meant to take fully flat vectors, and here you pass it the metric which acts on trees. You are not seeing issues because your tests do not use nested trees, but this will fail. Can you modify the logic so that it simply does everything using metric on the non-flat vectors? This may require changing a few tests on the function here and there but should be easy

**Comment by ismael-mendoza** [comment]
Thanks @AdrienCorenflos 

```
an you modify the logic so that it simply does everything using metric on the non-flat vectors?
 ```

just to make sure I understand, do you want me to remove `metric` from `_barker_sample_nd` entirely and only pass in flat vectors to this function? and the metric logic should live in `_barker_sample` ?

**Comment by AdrienCorenflos** [comment]
Well, I think you can just delete the nd version and do everything inside
the `_barker_sample`. It made sense when there was a pure flat logic, but
it's not the case anymore

**Comment by ismael-mendoza** [comment]
oh I think I understand your suggestion now, will modify and commit soon

**Comment by AdrienCorenflos** [comment]
Thanks, I'll check tomorrow. @junpenglao given I've touched that pr if you could also have a quick glance that would be useful

File: blackjax/mcmc/metrics.py
Comment (junpenglao) [inline blackjax/mcmc/metrics.py:69]: Let's turn this to required kwarg:
```suggestion
        *,
        inv: bool,
        trans: bool,
```

File: blackjax/mcmc/metrics.py
Comment (junpenglao) [inline blackjax/mcmc/metrics.py:236]: I think python if/else here also works as `inv` and `trans` are input as static parameters
```suggestion
        if inv:
            left_hand_side_matrix = inv_mass_matrix_sqrt
        else:
            left_hand_side_matrix = mass_matrix_sqrt
        if trans:
            left_hand_side_matrix = left_hand_side_matrix.T

        scaled = linear_map(left_hand_side_matrix, ravelled_element)
```

**Review by junpenglao — COMMENTED** [review]
Some minor nit, could you try to see if the suggested changes works?

**Comment by ismael-mendoza** [comment]
@junpenglao thanks for the suggestion, I think works but I see that one of the HMC benchmarks had a regression. Although it's unclear to me how that might have happened after just implementing your suggestion

**Comment by junpenglao** [comment]
> @junpenglao thanks for the suggestion, I think works but I see that one of the HMC benchmarks had a regression. Although it's unclear to me how that might have happened after just implementing your suggestion

Yeah the regression is not related, i will investigate.

**Comment by junpenglao** [comment]
@AdrienCorenflos any more comments? Otherwise LGTM.

**Comment by AdrienCorenflos** [comment]
Nope, I had no additional comments when I asked for your opinion :)

**Review by junpenglao — APPROVED** [review]

**Comment by junpenglao** [comment]
Thank you @ismael-mendoza !!

**Pull request merged**
